File: Aula 1/JS/5_polimorfismo_sobrescrita.js
// Classe pai
class Animal {
    constructor(nome, idade) {
        this.nome = nome;
        this.idade = idade;
    }

    fazBarulho() {
        console.log(`O animal ${this.nome} faz barulho`);
    }
}

// Subclasse
class Cachorro extends Animal {
    constructor(nome, idade, raca) {
        super(nome, idade);
        this.raca = raca;
    }

    fazBarulho() {
        console.log(`O cachorro ${this.nome} faz auau`);
    }
}

// Instancia
const animal1 = new Animal("Rex", 5);
const cachorro1 = new Cachorro("Toby", 10, "pug");

animal1.fazBarulho();
cachorro1.fazBarulho();

// // Subclasse
// class Gato extends Animal {
//     constructor(nome, idade, raca) {
//         super(nome, idade);
//         this.raca = raca;
//     }

//     fazBarulho() {
//         console.log(`O gato ${this.nome} faz miau`);
//     }
// }

// const gato1 = new Gato("Mingau", 2, "persa");
// gato1.fazBarulho();

File: Aula 1/C#/5_polimorfismo_sobrescrita.cs
namespace Course
{
    // Classe pai
    internal class Animal
    {
        public string Nome;
        public int Idade;

        public Animal(string nome, int idade)
        {
            Nome = nome;
            Idade = idade;
        }

        // Método virtual: pode ser sobrescrito pelas subclasses
        public virtual void FazBarulho()
        {
            Console.WriteLine($"O animal {Nome} faz barulho");
        }
    }

    // Subclasse
    internal class Cachorro : Animal
    {
        public string Raca;

        public Cachorro(string nome, int idade, string raca) : base(nome, idade)
        {
            Raca = raca;
        }

        // Sobrescrita: substitui o comportamento herdado
        public override void FazBarulho()
        {
            Console.WriteLine($"O cachorro {Nome} faz auau");
        }
    }

    internal class Program
    {
        static void Main(string[] args)
        {
            // Instâncias
            Animal animal1 = new Animal("Rex", 5);
            Cachorro cachorro1 = new Cachorro("Toby", 10, "pug");

            // Mesmo nome de método, comportamentos diferentes
            animal1.FazBarulho();
            cachorro1.FazBarulho();

            //internal class Gato : Animal { ... FazBarulho() → "faz miau" }
            //Gato gato1 = new Gato("Mingau", 2, "persa");
            //gato1.FazBarulho();
        }
    }
}
